src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -934,22 +931,36 @@ MergeTreeData::MutableDataPartPtr Fetcher::downloadPartToDisk(
 
     CurrentMetrics::Increment metric_increment{CurrentMetrics::ReplicatedFetch};
 
-    for (size_t i = 0; i < projections; ++i)
+    try
     {
-        String projection_name;
-        readStringBinary(projection_name, in);
-        MergeTreeData::DataPart::Checksums projection_checksum;
+        for (size_t i = 0; i < projections; ++i)
+        {
+            String projection_name;
+            readStringBinary(projection_name, in);
+            MergeTreeData::DataPart::Checksums projection_checksum;
+
+            auto projection_part_storage = data_part_storage->getProjection(projection_name + ".proj");
+            projection_part_storage->createDirectories();
+            downloadBaseOrProjectionPartToDisk(
+                replica_path, projection_part_storage, sync, in, projection_checksum, throttler);
+            checksums.addFile(
+                projection_name + ".proj", projection_checksum.getTotalSizeOnDisk(), projection_checksum.getTotalChecksumUInt128());
+        }
 
-        auto projection_part_storage = data_part_storage->getProjection(projection_name + ".proj");
-        projection_part_storage->createDirectories();
-        downloadBaseOrProjectionPartToDisk(
-            replica_path, projection_part_storage, sync, in, projection_checksum, throttler);
-        checksums.addFile(
-            projection_name + ".proj", projection_checksum.getTotalSizeOnDisk(), projection_checksum.getTotalChecksumUInt128());
+        // Download the base part
+        downloadBaseOrProjectionPartToDisk(replica_path, data_part_storage, sync, in, checksums, throttler);
+    }
+    catch (const Exception & e)
+    {
+        /// Remove the whole part directory if fetch of base
+        /// part or fetch of any projection was stopped.
+        if (e.code() == ErrorCodes::ABORTED)
+        {
+            data_part_storage->removeRecursive();
+            data_part_storage->commitTransaction();
+        }
+        throw;
```

Comment:
What about other error codes? See https://github.com/ClickHouse/ClickHouse/issues/54446

Reply:
This PR just restores the previous behaviour with one small fix for projections. This behaviour existed all the time. Probably we rely on background thread for deleting of temp directories.